### DISCUSSION TITLE
llama.cpp model tweaking

### DISCUSSION THREAD

**Discussion by Mugga6315** [skye-harris/hass_local_openai_llm#81] [General]
To continue our talk [https://github.com/skye-harris/hass_local_openai_llm/issues/80](url)

I named this one now, that we can maybe talk about models and tweaking. Also tried the qwen 3.5 9B model now with my 4060 TI (16gb). And results were quite good.

My setup in regard of HA is a little bit complicated because I was not 100% consistent with naming. So I would say i have 80% named in english together with adding alias for german and english. Normally I was also purely talking in english via voice satellites but with the Qwen 3.6 35B A3B I got pretty good results speaking in german. So this is what I also used in my testing.
The 9B MTP variant wasnt working, in regard of using german, he could for example not translate german word for bedroom lights.

Here are the numbers:
| Metric | 35B-A3B IQ3_XXS | 9B Q8 non-MTP |
|---|---|---|
| VRAM (measured) | 13,934 MiB | 8,948 MiB |
| Prefill — cold (~4k new tokens) | ~1,720–1,900 t/s | ~2,400–2,750 t/s |
| Prefill — larger (~6k new tokens) | ~1,910–1,970 t/s | ~2,700–2,840 t/s |
| Warm follow-up (cached) | ~0.26–0.58 s | ~0.6–3.3 s |
| Decode (generation) | ~95–133 t/s | ~29–37 t/s |
| MTP draft acceptance | ~1.0 | n/a (no MTP) |

The 9B was working pretty good so far for me, and also as you said output is totally fine because the prefill is much more important for voice usage in HA. 

My current config.yaml
```
macros:
  common_args: >
    --port ${PORT}
    --ctx-size 16843
    -ngl -1 --no-mmap --no-mmproj --parallel 1
    --reasoning off --flash-attn on --jinja --metrics --slots

models:
  qwen3.6-35b-a3b-mtp-homeassistant:
    cmd: |
      llama-server ${common_args}
      -hf unsloth/Qwen3.6-35B-A3B-MTP-GGUF:UD-IQ3_XXS
      --spec-type draft-mtp --spec-draft-n-max 2 --spec-draft-p-min 0.5
      --ctx-checkpoints 64
    ttl: 0
  qwen3.5-9b-mtp-homeassistant:
    cmd: |
      llama-server ${common_args}
      -hf unsloth/Qwen3.5-9B-MTP-GGUF:Q8_0
      --spec-type draft-mtp --spec-draft-n-max 2 --spec-draft-p-min 0.5
      --ctx-checkpoints 64
    ttl: 0
  qwen3.5-9b-homeassistant:
    cmd: |
      llama-server ${common_args}
      -hf unsloth/Qwen3.5-9B-GGUF:Q8_0
      --spec-type ngram-simple --spec-draft-n-max 64
      --ctx-checkpoints 64
    ttl: 0
``` 

Overall I'm quite happy with current setup. Only issues I current have a inconsistent answers (maybe because of 0.5 temperature). And I still cant get calendar events working, he always tells me there are no events even tough he as access to the calendar. Maybe I have to ditch the LLM blueprint I been using, do you have anything in regard of calendar activated as script or so?

Also I'm experimenting with Weaviate, have this one already running and wanted to use it to give certain context per questions. For example when I'm asking "Todays solar production" he just gives me kWh value which is normally totally fine, but I would like a summary, so kWh, together with highest peak in W and also current battery level. 

And to have more consistent guardrails about some tasks. For example when I was giving command "Turn on bedroom tv" he was also turning on the lights, which I never said.

**Comment by skye-harris** [comment]
Frankly, Im not liking Qwen3.5 for my voice assistant much. I don't like how overly verbose it is, and is resistant to prompting around some of its behaviours. I'm very tempted to just return to Qwen3VL 8B for the time being, although Gemma4 12B dense has dropped and might be worth trying out.

> Maybe I have to ditch the LLM blueprint I been using, do you have anything in regard of calendar activated as script or so?

I just exposed the calendars to Assist and they work, nothing special needed there I don't think?

> Also I'm experimenting with Weaviate, have this one already running and wanted to use it to give certain context per questions. For example when I'm asking "Todays solar production" he just gives me kWh value which is normally totally fine, but I would like a summary, so kWh, together with highest peak in W and also current battery level.

I need to get back to that some day... Ive got changes locally that implement a reranker model which definitely improves the results from it, but I've not really done any _proper_ testing or anything with it in a while now. I use it myself for a few contextual bits of info but it needs more time and thought put into it than Ive been able to for a while.

**Comment by Mugga6315** [comment]
Quite interesting, for me its the other way around: Gemma is "talking too much" and Qwen usually answers precisely. But maybe this is also a language thing and depend on your entities and stuff.

Gave the 12B a test today, and it wasnt working for me. Answer were not wrong, but do descriptive and repetitive. For me currently the Qwen3.6-35B-A3B seems to be the sweet spot.

But maybe in a week we have totally different models, wild times...

For time being I will also look into a few other ways to restructure and tidy my entities in HA, this will maybe also help HA. And see if I can get historical data working

**Comment by skye-harris** [comment]
Yeah for the meantime Ive settled back on trusty old Qwen3 VL

I have Gemma4 12B 4bit QAT quant but it ended up looping attempting to call the wrong tools. I'll give a Q6 quant a try at some point as that's normally where I draw the line when grabbing models, but wanted to see how the QAT variants went.

> And see if I can get historical data working

I did some testing in my other integration on this, had a little success but it was a bit finicky. This was _before_ adding the Home Control API stuff though, and tweaking some of the directives in there may assist on the entity history front as far as tool usage itself goes. The bigger issue I had was tho that the history context for sensors can be absolutely huge depending upon its activity granularity. For example, temperature sensors can change value by small amounts quite often, resulting in very large datasets when querying for eg the past 24 hours of data, and can result in several-thousands worth of tokens in history data for the LLM to then process, and haphazardly filtering the content can result in the actual data that the model is looking for being inadvertently removed.

It's not up to date with the latest changes for a few releases now but you can see this branch [here](https://github.com/skye-harris/llm_intents/tree/entity-history) and some discussions on it [here](https://github.com/skye-harris/llm_intents/discussions/87).

**Comment by Mugga6315** [comment]
I'm currently into fine tuning Qwen 3.5 4B model. Making I think some progress, but its a really tedious process. Getting all the training data right, is really not that easy. But I think this is also the reason, why we still didnt have a really good training LLM just for HA.
I'm thinking how I could workout a more automated process for creating, training and testing in HA to iterate faster and remove the tedious testing I'm doing.
Currently really tempted to just buy 2x 5070 Ti, use vllm and call it a day. But man this is much money and the prices are really annoying. If 5090 would be release price it would be a nobrainer to get this instead of dual 5070 ti, but with prices nowadays ...

**Reply by skye-harris** [reply]
Mmhmm _good_ training data can be really tedious. I've been fine-tuning piper voice models locally for a little while now, using voice lines & their english subtitles from Baldurs Gate. I have the benefit of a huge library of voice lines, pre-trimmed and transcribed, and curating this to actual good quality samples for fine-tuning is **such** a tedious chore! Much more fun than the stock piper voices though 😄

> I'm thinking how I could workout a more automated process for creating, training and testing in HA to iterate faster and remove the tedious testing I'm doing.

I've been playing with using Gemma4 to aid mine: testing out auto-tagging of samples for things like voice tone, emotion, and the like, as well as passing in 2 audio files and designating one the `gold sample` and the other the `test sample` to be evaluated. The results don't need to be "perfect", just "good enough", but I've not managed to get things working consistently well on that so far.

> Currently really tempted to just buy 2x 5070 Ti, use vllm and call it a day. But man this is much money and the prices are really annoying. If 5090 would be release price it would be a nobrainer to get this instead of dual 5070 ti, but with prices nowadays ...

Sounds like a slippery slope to me :) I've got a single 5060Ti and a second one is very tempting... but if Im sliding down that slippery slope then I just know Im not going to be happy with the (comparatively) slow VRAM of the dual 5060s

**Comment by Drizzt321** [comment]
I've built myself a test harness to test various models against my local GPU (haven't updated the reports yet). I think some of the scoring still needs some amount of improvement, but it's a repeatable method of testing with multiple sizes of HA entities, etc, matching as closely as I can to what HA actually sends. 

https://github.com/Drizzt321/ha-voiceagent-llm-benchmark

**Reply by skye-harris** [reply]
There's a somewhat official set of benchmarks/evals for Home Assistant [here](https://github.com/allenporter/home-assistant-datasets/), maintained by HA devs and contributors, that you may be interested in also :)

**Reply by Drizzt321** [reply]
I seem to have missed finding those, thanks!

**Reply by Mugga6315** [reply]
Will definitely have a look at this. I also invested quite some time and LLM power into building my own Benchmarking suite. It's in two parts splitted. A speed benchmark to find like best settings per model and backend vllm/llama.cpp. And a second part which tests quality for different domains and one is for HA with currently 70 test cases. Mixing German and English to test my like "special setup". Were I usually have everything named and setup in English in HA but talking in German to my assistant.
Once I polished this to a certain amount I can also publish this one.

Btw. I'm currently implementing this like "second brain" idea. First tests look promising. It's a extra integration which is intercepting into the assist pipeline so configurable to be used with this integration here and plus on top HA mcp server variants. It's surving two purposes giving HA assist direct access via mcp to also historical data for example. And second saving memory in a MD wiki. So when I say for example "remember my Wi-Fi password is 1234" he writes this into memory. Or also like rules "when the weather is sunny then say at the end grab your sunglasses".

**Reply by Drizzt321** [reply]
Yeah, I'm looking towards "second brain" type thing as well, but also an explicit knowledgebase. Currently implemented as a dir & .md based thing (see https://github.com/Drizzt321/ai-tools/tree/main/project-knowledgebase), but looking towards building an actual FTS/Vector Search/tagged/etc KB that will have both human an MCP interface/interactivity. 

Hopefully I can finish the base specs in the next few weeks, so I can actually push ahead with prototype/experiment stuff, and just get some experience/things working towards a more complete version.
